### PR TITLE
Bug 1949481: add DeepCopy to avoid SharedInformer cache mutation

### PR DESCRIPTION
**Pull request by gabemontero**
adjust fsm status update grouping to account for deepcopies
reduce logging

### CODE REVIEW

**Comment by gabemontero** [comment]
image eco tests passed by an alert firing with the cvo failed the suite

/test e2e-aws-image-ecosystem

**Comment by gabemontero** [comment]
sample upgraded fine in upgrade job, other stuff failed

/test e2e-aws-upgrade

**Comment by gabemontero** [comment]
last upgrade e2e only had 1 known sig-network error which is unrelated:  https://bugzilla.redhat.com/show_bug.cgi?id=1943334 and https://bugzilla.redhat.com/show_bug.cgi?id=1927264

/retest

**Comment by gabemontero** [comment]
@xiuwang for verification from QE my thought is follow the steps the customer did at https://access.redhat.com/documentation/en-us/red_hat_fuse/7.8/html/fuse_on_openshift_guide/get-started-admin#install-fuse-on-openshift4 and make sure no panics occur

disregard the log message chatter in the BZ wrt verification

**Review by adambkaplan — COMMENTED** [review]
A few questions and nits on logging.

File: pkg/stub/handler.go [hunk 1]
```diff
@@ -805,27 +826,25 @@ func (h *Handler) Handle(event util.Event) error {
...
-		if len(cfg.Spec.Architectures) == 0 {
-			cfg = h.updateCfgArch(cfg)
-		}
-
-		h.StoreCurrentValidConfig(cfg)
-
-		// now that we have stored the skip lists in status,
 		// cycle through the skip lists and update the managed flag if needed
-		for _, name := range cfg.Status.SkippedTemplates {
+		for _, name := range cfg.Spec.SkippedTemplates {
```
[the remark this answers is not in the document]
Comment (gabemontero) [reply pkg/stub/handler.go:830]: Had to do this as part of no longer grouping the status update with other changes.

Not grouping status update here with other changes helped with conflicts and mutations.

Semantically it is the same.  In hindsight, tracking "processed" config in the status may have been unnecessary, but we are where we are.

File: pkg/stub/handler.go [hunk 2]
```diff
@@ -805,27 +826,25 @@ func (h *Handler) Handle(event util.Event) error {
...
 			h.setSampleManagedLabelToFalse("template", name)
 		}
-		for _, name := range cfg.Status.SkippedImagestreams {
+		for _, name := range cfg.Spec.SkippedImagestreams {
```
[the remark this answers is not in the document]
Comment (gabemontero) [reply pkg/stub/handler.go:833]: Had to do this as part of no longer grouping the status update with other changes.

Not grouping status update here with other changes helped with conflicts and mutations.

Semantically it is the same.  In hindsight, tracking "processed" config in the status may have been unnecessary, but we are where we are.

File: pkg/stub/files.go [hunk 3]
```diff
@@ -16,7 +16,11 @@ import (
 func (h *Handler) processFiles(dir string, files []os.FileInfo, opcfg *v1.Config) error {
 	for _, file := range files {
 		if file.IsDir() {
-			logrus.Printf("processing subdir %s from dir %s", file.Name(), dir)
+			if opcfg.Status.Version != h.version {
+				logrus.Printf("processing subdir %s from dir %s", file.Name(), dir)
```
[the remark this answers is not in the document]
Comment (gabemontero) [reply pkg/stub/files.go:20]: that is discernible by other means ... I'm not inclined to change this at this point

File: pkg/stub/files.go [hunk 4]
```diff
@@ -28,7 +32,11 @@ func (h *Handler) processFiles(dir string, files []os.FileInfo, opcfg *v1.Config
 
 			continue
 		}
-		logrus.Printf("processing file %s from dir %s", file.Name(), dir)
+		if opcfg.Status.Version != h.version {
+			logrus.Printf("processing file %s from dir %s", file.Name(), dir)
```
[the remark this answers is not in the document]
Comment (gabemontero) [reply pkg/stub/files.go:36]: that is discernible by other means ... I'm not inclined to change this at this point

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Pull request merged**